compiler/src/iree/compiler/Dialect/Stream/Transforms/AutomaticReferenceCounting.cpp

```diff
@@ -0,0 +1,728 @@
+        indeterminateResources.insert(baseResource);
+      }
+    }
+    for (auto value : blockLiveness->out()) {
+      if (isa<IREE::Stream::ResourceType>(value.getType())) {
+        Value baseResource = lastUseSet.lookupResource(value);
+        LLVM_DEBUG({
+          llvm::dbgs() << "[arc] live-out resource ";
+          baseResource.printAsOperand(llvm::dbgs(), *asmState);
+          llvm::dbgs() << "; marking as indeterminate\n";
+        });
+        indeterminateResources.insert(baseResource);
+      }
+    }
+
+    for (auto &op : block) {
+      // TODO(benvanik): global analysis is required to know if calls are
+      // side-effecting. We could annotate the calls as LLVM does so that we
+      // could do local analysis and only pay attention to the operands/results.
+      // util.call supports tied operands but does not have a way to associate
+      // timepoints and this pass may never be able to work without that
+      // information. The most common case of calls today is in external modules
+      // not using `stream.cmd.call` and those are rare.
+      if (isa<CallOpInterface>(op)) {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "[arc] skipping function @" << getFuncName(funcOp)
+                   << " as it contains call ops (" << op.getName() << ")\n");
+        return;
+      }
+
+      // TODO(benvanik): broader analysis or at least constrained handling of
+      // scf. For now we bail if any ops from control flow dialects are found.
+      // TODO(benvanik): maybe only skip the block containing the ops.
+      if (op.getDialect()->getNamespace() == "scf") {
+        LLVM_DEBUG(llvm::dbgs()
+                   << "[arc] skipping function @" << getFuncName(funcOp)
+                   << " block ^bb" << blockIndex
+                   << " as it contains control flow ops (" << op.getName()
+                   << ")\n");
+        return;
+      }
+
+      // Special case ops that are not timeline-aware but interoperate.
+      if (auto immediateOp = dyn_cast<IREE::Stream::TimepointImmediateOp>(op)) {
+        coverage.add(immediateOp.getResultTimepoint());
+        continue;
+      } else if (auto importOp =
+                     dyn_cast<IREE::Stream::TimepointImportOp>(op)) {
+        coverage.add(importOp.getResultTimepoint());
+        continue;
+      }
+
+      // Bail on any resource which has lifetime management ops. They should not
+      // have been inserted yet and their presence likely indicates we've
+      // already run the pass on this input.
+      if (auto retainOp = dyn_cast<IREE::Stream::ResourceRetainOp>(op)) {
+        Value handledResource =
+            lastUseSet.lookupResource(retainOp.getOperand());
+        LLVM_DEBUG({
+          llvm::dbgs() << "[arc] existing retain of ";
+          handledResource.printAsOperand(llvm::dbgs(), *asmState);
+          llvm::dbgs() << "; marking as handled\n";
+        });
+        handledResources.insert(handledResource);
+        continue;
```

Comment:
When you say bail, do you want to just stop the analysis, or just ignore the operation. Cause this is just doing the latter for now.

Reply:
Just ignore the particular resource the operation is acting on - we can still process other resources. Added to the comment to clarify that it's not trying to bail on the analysis.